chapter_preliminaries/autograd.ipynb

```diff
@@ -0,0 +1,445 @@
+    "outside of the recorded computational graph.\n",
+    "For example, say that `y` was calculated as a function of `x`,\n",
+    "and that subsequently `z` was calculated as a function of both `y` and `x`.\n",
+    "Now, imagine that we wanted to calculate\n",
+    "the gradient of `z` with respect to `x`,\n",
+    "but wanted for some reason to treat `y` as a constant,\n",
+    "and only take into account the role\n",
+    "that `x` played after `y` was calculated.\n",
+    "\n",
+    "Here, we can detach `y` using `stopGradient` to return \n",
+    "a new variable `u`\n",
+    "that has the same value as `y` but discards any information\n",
+    "about how `y` was computed in the computational graph.\n",
+    "In other words, the gradient will not flow backwards through `u` to `x`.\n",
+    "Thus, the following backpropagation function computes\n",
+    "the partial derivative of `z = u * x` with respect to `x` while treating `u` as a constant,\n",
+    "instead of the partial derivative of `z = x * x * x` with respect to `x`.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "try (GradientCollector gc = Engine.getInstance().newGradientCollector()) {\n",
+    "    NDArray y = x.mul(x);\n",
+    "    NDArray u = y.stopGradient();\n",
+    "    NDArray z = u.mul(x);\n",
+    "    gc.backward(z);\n",
+    "    System.out.println(x.getGradient().eq(u));\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 36
+   },
+   "source": [
+    "We can subsequently invoke backpropagation on `y` to get the derivative of `y = x * x` with respect to `x`, which is `2 * x`.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "try (GradientCollector gc = Engine.getInstance().newGradientCollector()) {\n",
+    "    NDArray y = x.mul(x);\n",
+    "    y = x.mul(x);\n",
+    "    gc.backward(y);\n",
+    "    System.out.println(x.getGradient().eq(x.mul(2)));\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 40
+   },
+   "source": [
+    "## Computing the Gradient of Java Control Flow\n",
+    "\n",
+    "One benefit of using automatic differentiation\n",
+    "is that even if building the computational graph of a function\n",
+    "required passing through a maze of Java control flow\n",
+    "(e.g., conditionals, loops, and arbitrary function calls),\n",
+    "we can still calculate the gradient of the resulting variable.\n",
+    "In the following snippet, note that\n",
+    "the number of iterations of the `while` loop\n",
+    "and the evaluation of the `if` statement\n",
+    "both depend on the value of the input `a`.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public NDArray l2Norm(NDArray w){\n",
```

Comment:
you should be able to use norm op with 0.10.0 snapshot

Reply:
Ahh yes. When I did this I don't think norm operator was merged yet to master. Will update it now.

Reply:
Just checked and it seems it was merged when I was developing so I used it but forgot to remove the l2norm() function declaration here so I'll do that.